arrow-cpp fails to build or install on Darwin

**Describe the bug**
arrow-cpp cannot be installed on Darwin

**To Reproduce**
Steps to reproduce the behavior:
1. nix-env -i arrow-cpp

**Expected behavior**
Successful installation

**Additional context**

nix-build -A arrow-cpp from nixpkgs master fails with this error:

```
CMake Error at /nix/store/mvj3a6zmkx12kv66jx4fcxxmqsb632g9-zstd-1.4.5-dev/lib/cmake/zstd/zstdTargets.cmake:75 (message):
  The imported target "zstd::libzstd_static" references the file

     "/nix/store/35q938v1xv31c9ph8p4vqcf81rff0q89-zstd-1.4.5/lib/libzstd.a"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/nix/store/mvj3a6zmkx12kv66jx4fcxxmqsb632g9-zstd-1.4.5-dev/lib/cmake/zstd/zstdTargets.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /nix/store/mvj3a6zmkx12kv66jx4fcxxmqsb632g9-zstd-1.4.5-dev/lib/cmake/zstd/ZSTDConfig.cmake:1 (include)
  cmake_modules/ThirdpartyToolchain.cmake:156 (find_package)
  cmake_modules/ThirdpartyToolchain.cmake:1857 (resolve_dependency)
  CMakeLists.txt:458 (include)
```

**Notify maintainers**
@veprbl 

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

 - system: `"x86_64-darwin"`
 - host os: `Darwin 19.5.0, macOS 10.15.5`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.7`
 - channels(lclaude): `"nixpkgs-20.09pre233849.1d801806827"`
 - nixpkgs: `/Users/lclaude/.nix-defexpr/channels/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

--- a/pkgs/development/libraries/arrow-cpp/default.nix
+++ b/pkgs/development/libraries/arrow-cpp/default.nix
@@ -42,6 +42,13 @@ in stdenv.mkDerivation rec {
       sha256 = "1nl4y1rwdl0gn67v7l05ibc4lwkn6x7fhwbmslmm08cqmwfjsx3y";
       stripLen = 1;
     })
+
+    # fix build for "ZSTD_SOURCE=SYSTEM"
+    (fetchpatch {
+      url = "https://github.com/apache/arrow/commit/13cb3dbded1928d2e96574895bebaf9098a4796d.diff";
+      sha256 = "12z3ys47qp2x8f63lggiyj4xs2kmg804ri4xqysw5krbjz2hr6rb";
+      stripLen = 1;
+    })
   ] ++ lib.optionals (!enableShared) [
     # The shared jemalloc lib is unused and breaks in static mode due to missing -fpic.
     ./jemalloc-disable-shared.patch
@@ -89,6 +96,7 @@ in stdenv.mkDerivation rec {
     "-DARROW_WITH_SNAPPY=ON"
     "-DARROW_WITH_ZLIB=ON"
     "-DARROW_WITH_ZSTD=ON"
+    "-DARROW_ZSTD_USE_SHARED=OFF" # TODO use shared zstd once #91984 hits the master
     # Parquet options:
     "-DARROW_PARQUET=ON"
     "-DPARQUET_BUILD_EXECUTABLES=ON"

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -11426,6 +11426,8 @@ in
   arrayfire = callPackage ../development/libraries/arrayfire {};
 
   arrow-cpp = callPackage ../development/libraries/arrow-cpp ({
+    # Workaround until #91984 is in, this will need to be removed later.
+    zstd = zstd.override { enableShared = false; };
   } // stdenv.lib.optionalAttrs (stdenv.hostPlatform.isi686 && stdenv.cc.isGNU) {
     stdenv = overrideCC stdenv buildPackages.gcc6; # hidden symbol `__divmoddi4'
   });

